flaky test: k8s.io/kubernetes/pkg/controller/volume/attachdetach: reconciler

### Which jobs are flaking?

pull-kubernetes-unit
https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-unit

### Which tests are flaking?

k8s.io/kubernetes/pkg/controller/volume/attachdetach: reconciler Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce

### Since when has it been flaking?

unknown, maybe recently?


### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit

### Reason for failure (if possible)

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121674/pull-kubernetes-unit/1720011568269234176

```
=== FAIL: pkg/controller/volume/attachdetach/reconciler Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce (0.65s)
W1102 09:49:25.408994   41304 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
    reconciler_test.go:1534: Warning: Volume <fake-plugin/volume-name> is not added to node  <fail-detach-node> yet. Will retry.
    reconciler.go:349: I1102 09:49:25.409555] Starting attacherDetacher.AttachVolume volume={"MultiAttachErrorReported":false,"VolumeName":"fake-plugin/volume-name","VolumeSpec":{"Volume":{"name":"volume-name","gcePersistentDisk":{"pdName":"volume-name","fsType":"fake"}},"PersistentVolume":{"metadata":{"creationTimestamp":null},"spec":{"accessModes":["ReadWriteOnce"]},"status":{}},"ReadOnly":false,"InlineVolumeSpecForCSIMigration":false,"Migrated":false},"NodeName":"fail-detach-node","ScheduledPods":[{"metadata":{"name":"pod-uid1","namespace":"pod-uid1","uid":"pod-uid1","creationTimestamp":null},"spec":{"containers":null},"status":{}}]}
    reconciler.go:352: I1102 09:49:25.409786] attacherDetacher.AttachVolume started volumeName="fake-plugin/volume-name" nodeName="fail-detach-node" scheduledPods=["pod-uid1/pod-uid1"]
I1102 09:49:25.409966   41304 operation_generator.go:400] AttachVolume.Attach succeeded for volume "volume-name" (UniqueName: "fake-plugin/volume-name") from node "fail-detach-node" 
    actual_state_of_world.go:512: I1102 09:49:25.410057] Add new node to nodesToUpdateStatusFor node="fail-detach-node"
    actual_state_of_world.go:520: I1102 09:49:25.410158] Report volume as attached to node node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler_test.go:1603: Volume <fake-plugin/volume-name> is attached to node <fail-detach-node>: Attached
    reconciler_test.go:1626: Volume <fake-plugin/volume-name> is reported as attached to node <fail-detach-node>: true
    actual_state_of_world.go:440: I1102 09:49:25.971598] Set detach request time to current time for volume on node node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:25.971784] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:25.971918] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:229: I1102 09:49:25.971997] Cannot detach volume because it is still mounted node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:25.982363] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:25.982628] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:229: I1102 09:49:25.982795] Cannot detach volume because it is still mounted node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:25.993214] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:25.993350] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:229: I1102 09:49:25.993447] Cannot detach volume because it is still mounted node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:26.003863] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:26.004014] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:229: I1102 09:49:26.004150] Cannot detach volume because it is still mounted node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:26.014612] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:26.014768] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:229: I1102 09:49:26.014864] Cannot detach volume because it is still mounted node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:215: E1102 09:49:26.061227] Failed to get health of node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:223: E1102 09:49:26.061359] Failed to get taint specs for node err="node \"fail-detach-node\" not found" node="fail-detach-node"
    reconciler.go:259: I1102 09:49:26.061450] Starting attacherDetacher.DetachVolume node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler.go:271: I1102 09:49:26.061670] attacherDetacher.DetachVolume started: this volume is not safe to detach, but maxWaitForUnmountDuration expired, force detaching duration="50ms" node="fail-detach-node" volumeName="fake-plugin/volume-name"
    reconciler_test.go:1597: Check volume <fake-plugin/volume-name> is attached to node <fail-detach-node>, got Attached, expected Uncertain
    actual_state_of_world.go:360: I1102 09:49:26.061846] Volume is already added to attachedVolume list to node, update device path volumeName="fake-plugin/volume-name" node="fail-detach-node" devicePath=""
    actual_state_of_world.go:376: I1102 09:49:26.061997] Volume is already added to attachedVolume list to the node volumeName="fake-plugin/volume-name" node="fail-detach-node" currentAttachState=false
DONE 49090 tests, 25 skipped, 1 failure in 2.136s
+++ [1102 10:17:50] prune-junit-xml not found; installing from hack/tools
processing junit xml file : /logs/artifacts/junit_20231102-093744.xml
done.
+++ [1102 10:17:50] Saved JUnit XML test report to /logs/artifacts/junit_20231102-093744.xml
make: *** [Makefile:184: test] Error 1

```


### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig storage

--- a/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go
+++ b/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go
@@ -630,9 +630,7 @@ func Test_Run_OneVolumeAttachAndDetachUncertainNodesWithReadWriteOnce(t *testing
 	if podAddErr != nil {
 		t.Fatalf("AddPod failed. Expected: <no error> Actual: <%v>", podAddErr)
 	}
-	waitForVolumeAttachedToNode(t, generatedVolumeName, nodeName2, asw)
-	verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName2, cache.AttachStateAttached, asw)
-
+	waitForVolumeAttachStateToNode(t, generatedVolumeName, nodeName2, cache.AttachStateAttached, asw)
 }
 
 func Test_Run_UpdateNodeStatusFailBeforeOneVolumeDetachNodeWithReadWriteOnce(t *testing.T) {
@@ -747,13 +745,7 @@ func Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce(t *testing.T) {
 	// Delete the pod, but detach will fail
 	dsw.DeletePod(types.UniquePodName(podName1), generatedVolumeName, nodeName1)
 
-	// The first detach will be triggered after at least 50ms (maxWaitForUnmountDuration in test).
-	// Right before detach operation is performed, the volume will be first removed from being reported
-	// as attached on node status (RemoveVolumeFromReportAsAttached). After detach operation which is expected to fail,
-	// controller then treats the attachment as Uncertain.
-	// Here it sleeps 100ms so that detach should be triggered already at this point.
-	time.Sleep(100 * time.Millisecond)
-	verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName1, cache.AttachStateUncertain, asw)
+	waitForVolumeAttachStateToNode(t, generatedVolumeName, nodeName1, cache.AttachStateUncertain, asw)
 	verifyVolumeReportedAsAttachedToNode(t, logger, generatedVolumeName, nodeName1, false, asw, volumeAttachedCheckTimeout)
 
 	// Add a second pod which tries to attach the volume to the same node.
@@ -843,9 +835,7 @@ func Test_Run_OneVolumeAttachAndDetachTimeoutNodesWithReadWriteOnce(t *testing.T
 	if podAddErr != nil {
 		t.Fatalf("AddPod failed. Expected: <no error> Actual: <%v>", podAddErr)
 	}
-	waitForVolumeAttachedToNode(t, generatedVolumeName, nodeName2, asw)
-	verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName2, cache.AttachStateAttached, asw)
-
+	waitForVolumeAttachStateToNode(t, generatedVolumeName, nodeName2, cache.AttachStateAttached, asw)
 }
 
 // Populates desiredStateOfWorld cache with one node/volume/pod tuple.
@@ -1485,35 +1475,33 @@ func verifyNewAttacherCallCount(
 	}
 }
 
-func waitForVolumeAttachedToNode(
+func waitForVolumeAttachStateToNode(
 	t *testing.T,
 	volumeName v1.UniqueVolumeName,
 	nodeName k8stypes.NodeName,
+	expectedAttachState cache.AttachState,
 	asw cache.ActualStateOfWorld) {
 
 	err := retryWithExponentialBackOff(
 		time.Duration(500*time.Millisecond),
 		func() (bool, error) {
 			attachState := asw.GetAttachState(volumeName, nodeName)
-			if attachState == cache.AttachStateAttached {
+			if attachState == expectedAttachState {
 				return true, nil
 			}
-			t.Logf(
-				"Warning: Volume <%v> is not attached to node  <%v> yet. Will retry.",
-				volumeName,
-				nodeName)
-
+			t.Logf("Warning: expected attach state: %v, actual attach state: %v. Will retry.",
+				expectedAttachState, attachState)
 			return false, nil
 		},
 	)
 
 	attachState := asw.GetAttachState(volumeName, nodeName)
-	if err != nil && attachState != cache.AttachStateAttached {
-		t.Fatalf(
-			"Volume <%v> is not attached to node  <%v>.",
-			volumeName,
-			nodeName)
+	if err != nil && attachState != expectedAttachState {
+		t.Fatalf("Volume <%v> is not in expected attach state: %v, actual attach state: %v",
+			volumeName, expectedAttachState, attachState)
 	}
+
+	t.Logf("Volume <%v> is attached to node <%v>: %v", volumeName, nodeName, expectedAttachState)
 }
 
 func waitForVolumeAddedToNode(

